// enzyme-to-json/build/mount.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mountToShallowJson = exports.mountToDeepJson = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _lodash = require('lodash.omitby');

var _lodash2 = _interopRequireDefault(_lodash);

var _lodash3 = require('lodash.isnil');

var _lodash4 = _interopRequireDefault(_lodash3);

var _objectValues = require('object-values');

var _objectValues2 = _interopRequireDefault(_objectValues);

var _reactCompat = require('enzyme/build/react-compat');

var _Utils = require('enzyme/build/Utils');

var _Debug = require('enzyme/build/Debug');

var _ShallowTraversal = require('enzyme/build/ShallowTraversal');

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function instToJson(inst, options) {
  var mode = options.mode;


  if (typeof inst === 'string' || typeof inst === 'number') {
    return inst;
  }
  if (!inst) {
    return null;
  }

  if (inst._stringText || mode === 'shallow' && typeof inst._stringText === 'string') {
    return inst._stringText;
  }

  if (!inst.getPublicInstance) {
    var internal = (0, _Utils.internalInstance)(inst);
    return instToJson(internal, options);
  }
  var publicInst = inst.getPublicInstance();

  if (typeof publicInst === 'string' || typeof publicInst === 'number') {
    return publicInst;
  }
  if (!publicInst && !inst._renderedComponent) {
    return null;
  }

  var currentElement = inst._currentElement;
  var type = (0, _Debug.typeName)(currentElement);
  var props = (0, _lodash2.default)(_extends({}, (0, _Utils.propsOfNode)(currentElement)), mode !== 'normal' ? _utils.omitFromPropsCompatible : _utils.omitFromPropsMinimal);

  if (options.noKey !== true && !(0, _lodash4.default)(currentElement.key)) {
    props.key = currentElement.key;
  }

  var children = [];
  if ((0, _reactCompat.isDOMComponent)(publicInst)) {
    var renderedChildren = inst._renderedChildren;
    if ((0, _lodash4.default)(renderedChildren)) {
      children.push.apply(children, _toConsumableArray((0, _ShallowTraversal.childrenOfNode)(currentElement)));
    } else {
      children.push.apply(children, _toConsumableArray((0, _objectValues2.default)(renderedChildren)));
    }
  } else if ((0, _reactCompat.isElement)(currentElement) && typeof currentElement.type === 'function') {
    if (mode === 'normal') {
      children.push(inst._renderedComponent);
    } else if (mode === 'deep') {
      // A component returns at most one element in React 15 and earlier.
      return instToJson(inst._renderedComponent, options);
    }
    // else if shallow, children are still an empty array
  }

  var childrenArray = children.map(function (n) {
    return instToJson(n, options);
  }).filter(mode === 'shallow' ? _utils.includeInChildrenCompatible : _utils.includeInChildrenMinimal);

  return {
    type,
    props,
    children: childrenArray.length ? childrenArray : null,
    $$typeof: Symbol.for('react.test.json')
  };
}

var wrapperToJson = function wrapperToJson(wrapper, options) {
  return wrapper.length > 1 ? wrapper.nodes.map(function (node) {
    return instToJson(node, options);
  }) : instToJson(wrapper.node, options);
};

var mountToDeepJson = function mountToDeepJson(wrapper) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return wrapperToJson(wrapper, _extends({
    mode: 'deep'
  }, options));
};

var mountToShallowJson = function mountToShallowJson(wrapper) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return wrapperToJson(wrapper, _extends({
    mode: 'shallow'
  }, options));
};

var mountToJson = function mountToJson(wrapper) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return wrapperToJson(wrapper, _extends({
    mode: 'normal'
  }, options));
};

exports.mountToDeepJson = mountToDeepJson;
exports.mountToShallowJson = mountToShallowJson;
exports.default = mountToJson;
